### PR TITLE
#113: Added unified error codes

### LINKED ISSUE #113: Use unified error codes
## Acceptance Criteria

- Error messages contain code tag, cause and mitigation
- Uses [error-reporting-java](https://github.com/exasol/error-reporting-java) library

### PR DESCRIPTION
**Pull request by morazow**
Fixes #113

### CODE REVIEW

File: src/main/scala/com/exasol/cloudetl/bucket/AzureBlobBucket.scala [hunk 1]
```diff
@@ -86,7 +87,14 @@ final case class AzureBlobBucket(path: String, params: StorageProperties) extend
   private[this] def regexParsePath(path: String): AccountAndContainer = path match {
     case AZURE_BLOB_PATH_REGEX(containerName, accountName, _) =>
       AccountAndContainer(accountName, containerName)
-    case _ => throw new BucketValidationException(s"Invalid Azure blob wasb(s) path: $path!")
+    case _ =>
+      throw new BucketValidationException(
+        ExaError
+          .messageBuilder("E-CSE-19")
+          .message("Azure blob wasb(s) path {{PATH}} is not valid.")
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/bucket/AzureBlobBucket.scala:94]: is wasb(s) a domain specific term?

File: src/main/scala/com/exasol/cloudetl/bucket/Bucket.scala [hunk 2]
```diff
@@ -54,7 +55,12 @@ abstract class Bucket extends LazyLogging {
     getRequiredProperties().foreach { key =>
       if (!properties.containsKey(key)) {
         throw new IllegalArgumentException(
-          s"Please provide a value for the $key property!"
+          ExaError
+            .messageBuilder("E-CSE-2")
+            .message("Required {{KEY}} property value is missing.")
+            .parameter("KEY", key)
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/bucket/Bucket.scala:61]: you can also do:
```suggestion
            .message("Required {{KEY}} property value is missing.", key)
```

File: src/main/scala/com/exasol/cloudetl/helper/ExasolColumnValueProvider.scala [hunk 3]
```diff
@@ -41,7 +48,13 @@ final case class ExasolColumnValueProvider(iterator: ExaIterator) extends JavaCl
     } else {
       val updatedBigDecimal = bigDecimal.setScale(scale, RoundingMode.HALF_UP)
       if (updatedBigDecimal.precision > precision) {
-        throw new IllegalArgumentException(s"Actual precision of big decimal value exceeds configured '$precision'.")
+        throw new IllegalArgumentException(
+          ExaError
+            .messageBuilder("E-CSE-24")
+            .message("Actual precision of big decimal value exceeds configured {{PRECISION}}.")
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/helper/ExasolColumnValueProvider.scala:54]: You could maybe add the value to make it easier to find that record. And in addition add the actual precision.

File: src/main/scala/com/exasol/cloudetl/helper/ParquetSchemaConverter.scala [hunk 4]
```diff
@@ -96,7 +97,14 @@ final case class ParquetSchemaConverter(isLowercaseSchemaEnabled: Boolean) exten
       case `jBoolean`      => getPrimitiveType(columnName, BOOLEAN, repetition, None, None)
       case `jSqlDate`      => getPrimitiveType(columnName, INT32, repetition, None, Option(dateType()))
       case `jSqlTimestamp` => getPrimitiveType(columnName, INT96, repetition, None, None)
-      case _ => throw new IllegalArgumentException(s"Cannot convert Exasol type '$columnType' to Parquet type.")
+      case _ =>
+        throw new IllegalArgumentException(
+          ExaError
+            .messageBuilder("E-CSE-22")
+            .message("Cannot convert Exasol type {{TYPE}} to Parquet type.")
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/helper/ParquetSchemaConverter.scala:104]: is that something we expect to happen?
If yes:
Add a mitiogation.
If no:
Maybe you can change type to F- and add `.ticketMitigation()`

File: src/main/scala/com/exasol/cloudetl/orc/converter/OrcConverterFactory.scala [hunk 5]
```diff
@@ -40,13 +40,15 @@ object OrcConverterFactory {
       case Category.VARCHAR   => StringConverter
       case _ =>
         throw new IllegalArgumentException(
-          s"Found orc unsupported type, '${orcType.getCategory}'."
+          ExaError
+            .messageBuilder("E-CSE-10")
+            .message("Orc primitive type {{PRIMITIVE_TYPE}} is not supported.")
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/orc/converter/OrcConverterFactory.scala:45]: same.

File: src/main/scala/com/exasol/cloudetl/orc/converter/OrcConverterFactory.scala [hunk 6]
```diff
@@ -59,7 +61,12 @@ object OrcConverterFactory {
       case Category.UNION  => UnionConverter(orcType)
       case _ =>
         throw new IllegalArgumentException(
-          s"Found orc unsupported type, '${orcType.getCategory}'."
+          ExaError
+            .messageBuilder("E-CSE-11")
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/orc/converter/OrcConverterFactory.scala:65]: same

File: src/main/scala/com/exasol/cloudetl/parquet/RowWriteSupport.scala [hunk 7]
```diff
@@ -200,7 +208,11 @@ class RowWriteSupport(schema: MessageType) extends WriteSupport[Row] {
           decimalBuffer
         } else {
           throw new IllegalStateException(
-            s"The precision $precision is too small for decimal value."
+            ExaError
+              .messageBuilder("E-CSE-9")
+              .message("The precision {{PRECISION}} is too small for decimal value.")
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/parquet/RowWriteSupport.scala:213]: can you add a mitigation?

File: src/main/scala/com/exasol/cloudetl/scriptclasses/ImportPath.scala [hunk 8]
```diff
@@ -2,12 +2,17 @@ package com.exasol.cloudetl.scriptclasses
...
+import com.exasol.errorreporting.ExaError
 
 object ImportPath {
 
   def generateSqlForImportSpec(meta: ExaMetadata, spec: ExaImportSpecification): String =
     throw new IllegalArgumentException(
-      "This script class is deprecated. Please use the FilesImportQueryGenerator " +
-        "class name. You can check the user guide for updated deployment scripts."
+      ExaError
+        .messageBuilder("W-CSE-8")
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/scriptclasses/ImportPath.scala:12]: Why is it a warning. I's in an exception, so won't it break the execution?

File: src/main/scala/com/exasol/cloudetl/storage/FileFormat.scala [hunk 9]
```diff
@@ -16,7 +18,16 @@ object FileFormat {
...
     case "PARQUET" => PARQUET
-    case _         => throw new IllegalArgumentException(s"Unsupported file format $fileFormat!")
+    case _ =>
+      throw new IllegalArgumentException(
+        ExaError
+          .messageBuilder("E-CSE-17")
+          .message("Provided file format {{FORMAT}} is not supported.")
+          .parameter("FORMAT", fileFormat)
+          .mitigation("Please use one of supported formats.")
+          .mitigation("You can check user guide for supported list of formats.")
```
Comment (jakobbraun) [inline src/main/scala/com/exasol/cloudetl/storage/FileFormat.scala:28]: maybe you can add a link to the user guide.
Also for others.

**Review by jakobbraun — CHANGES_REQUESTED** [review]

File: src/main/scala/com/exasol/cloudetl/bucket/Bucket.scala (hunk 2 above)
Comment (morazow) [reply src/main/scala/com/exasol/cloudetl/bucket/Bucket.scala:61]: This is nice, definitely more nicer for some of them.

File: src/main/scala/com/exasol/cloudetl/bucket/AzureBlobBucket.scala (hunk 1 above)
Comment (morazow) [reply src/main/scala/com/exasol/cloudetl/bucket/AzureBlobBucket.scala:94]: Yes, it is path scheme for Azure blob storage. Maybe I can reword it and remove this term.

**Pull request merged**